packages/connectivity/src/scp-cf/environment-accessor/xsuaa.ts

```diff
@@ -0,0 +1,102 @@
+import { createLogger } from '@sap-cloud-sdk/util';
+import { JwtPayload } from '../jsonwebtoken-type';
+import { decodeJwt } from '../jwt';
+import { XsuaaServiceCredentials } from './environment-accessor-types';
+import {
+  getServiceCredentialsList,
+  matchesAudience,
+  matchesClientId
+} from './env';
+
+const logger = createLogger({
+  package: 'connectivity',
+  messageContext: 'environment-accessor'
+});
+
+/**
+ * @internal
+ * Takes a decoded JWT and uses the client_id and audience claims to determine the XSUAA service instance
+ * that issued the JWT. Returns the credentials if a match is found, otherwise throws an error.
+ * If no decoded JWT is specified, then returns the first existing XSUAA credential service plan "application".
+ * @param token - Either an encoded or decoded JWT.
+ * @returns The credentials for a match, otherwise `null`.
+ */
+export function getXsuaaServiceCredentials(
+  token?: JwtPayload | string
+): XsuaaServiceCredentials {
+  const xsuaaCredentials = getServiceCredentialsList('xsuaa');
+
+  if (!xsuaaCredentials.length) {
+    throw Error(
+      'Could not find binding to the XSUAA service, that includes credentials.'
+    );
+  }
+
+  if (token) {
+    return getCredentialsWithJwt(
+      xsuaaCredentials,
+      typeof token === 'string' ? decodeJwt(token) : token
+    );
+  }
+  logger.debug('No JWT given to select XSUAA instance.');
+
+  return getCredentialsWithoutJwt(xsuaaCredentials);
+}
+
+function getCredentialsWithJwt(
+  xsuaaCredentials: XsuaaServiceCredentials[],
+  token: JwtPayload
+): XsuaaServiceCredentials {
+  const eligibleCredentials = xsuaaCredentials.filter(
+    credentials =>
+      matchesClientId(credentials, token) || matchesAudience(credentials, token)
+  );
+
+  logResult(eligibleCredentials, true);
+
+  if (eligibleCredentials.length) {
+    return eligibleCredentials[0];
+  }
+
+  logger.debug(
+    'Could not match XSUAA service instance using JWT client id or audience.'
+  );
+
+  return getCredentialsWithoutJwt(xsuaaCredentials);
+}
+
+function getCredentialsWithoutJwt(
+  xsuaaCredentials: XsuaaServiceCredentials[]
+): XsuaaServiceCredentials {
+  logResult(xsuaaCredentials, false);
+  return xsuaaCredentials[0];
+}
+
+function logResult(credentials: XsuaaServiceCredentials[], usedToken: boolean) {
```

Comment:
Perhaps a more speaking name could be used for this utility function

Reply:
I'll do another refactoring PR and I will think about it in the next one than. Hope that's ok.